Kernel/Graphics/VirtIOGPU/VirtIOGPU.cpp

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2021, Sahan Fernando <sahan.h.fernando@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/Graphics/VirtIOGPU/VirtIOGPU.h>
+#include <LibC/sys/ioctl_numbers.h>
+
+#define DEVICE_EVENTS_READ 0x0
+#define DEVICE_EVENTS_CLEAR 0x4
+#define DEVICE_NUM_SCANOUTS 0x8
+
+namespace Kernel::Graphics {
+
+VirtIOGPU::VirtIOGPU(PCI::Address address)
+    : VirtIODevice(address, "VirtIOGPU")
+    , m_scratch_space(MM.allocate_contiguous_kernel_region(8 * PAGE_SIZE, "VirtGPU Scratch Space", Region::Access::Read | Region::Access::Write))
+{
+    VERIFY(!!m_scratch_space);
+    dbgln("Gonna configure now");
+    if (auto cfg = get_config(ConfigurationType::Device)) {
+        m_device_configuration = cfg;
+        bool success = negotiate_features([&](u64 supported_features) {
+            u64 negotiated = 0;
+            if (is_feature_set(supported_features, VIRTIO_GPU_F_VIRGL))
+                dbgln("VirtIOGPU: VIRGL is not yet supported!");
+            if (is_feature_set(supported_features, VIRTIO_GPU_F_EDID))
+                dbgln("VirtIOGPU: EDID is not yet supported!");
+            return negotiated;
+        });
+        if (success) {
+            read_config_atomic([&]() {
+                m_num_scanouts = config_read32(*cfg, 0x8);
+            });
+            dbgln("VirtIOGPU: num_scanouts: {}", m_num_scanouts);
+            success = setup_queues(2); // CONTROLQ + CURSORQ
+        }
+        if (success) {
+            finish_init();
+        }
+        // 1. Get display information using VIRTIO_GPU_CMD_GET_DISPLAY_INFO
+        query_display_information();
+        // 2. Create BUFFER using VIRTIO_GPU_CMD_RESOURCE_CREATE_2D
+        create_framebuffer();
+        // 3. Attach backing storage using  VIRTIO_GPU_CMD_RESOURCE_ATTACH_BACKING
+        attach_backing_storage();
+        // 4. Use VIRTIO_GPU_CMD_SET_SCANOUT to link the framebuffer to a display scanout.
+        attach_framebuffer_to_scanout();
+        // 5. Render our test pattern
+        draw_test_pattern();
```

Comment:
should we be doing this normally?

Reply:
I'd like to leave that in, since it allows us to easily see a bit more about whether we were capable of drawing to the framebuffer at all before it crashed. That being said, maybe we should rename this to something like `draw_serenity_startup_image`, and change it from the japanese flag to something a bit more production oriented.

Reply:
that sounds good :)